domain/permission/service/service.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/permission"
+)
+
+// State describes retrieval and persistence methods for user permission on
+// various targets.
+type State interface {
+	// CreatePermission gives the user access per the provided spec.
+	// It requires the user/target combination has not already been
+	// created.
+	CreatePermission(ctx context.Context, spec UserAccessSpec) (permission.UserAccess, error)
+
+	// DeletePermission removes the given subject's (user) access to the
+	// given target.
+	DeletePermission(ctx context.Context, subject string, target permission.ID) error
+
+	// UpdatePermission updates the subject's (user) access to the given
+	// target to the given access.
+	UpdatePermission(ctx context.Context, subject string, target permission.ID, access permission.Access) error
+
+	// ReadUserAccessForTarget returns the subject's (user) access for the
+	// given user on the given target.
+	ReadUserAccessForTarget(ctx context.Context, subject string, target permission.ID) (permission.UserAccess, error)
+
+	// ReadUserAccessLevelForTarget returns the subject's (user) access level
+	// for the given user on the given target.
+	ReadUserAccessLevelForTarget(ctx context.Context, subject string, target permission.ID) (permission.Access, error)
+
+	// ReadAllUserAccessForUser returns a slice of the user access the given
+	// subject's (user) has for any access type.
+	ReadAllUserAccessForUser(ctx context.Context, subject string) ([]permission.UserAccess, error)
+
+	// ReadAllUserAccessForTarget return a slice of user access for all users
+	// with access to the given target.
+	ReadAllUserAccessForTarget(ctx context.Context, target permission.ID) ([]permission.UserAccess, error)
+
+	// ReadAllAccessTypeForUser return a slice of user access for the subject
+	// (user) specified and of the given access type.
+	// E.G. All clouds the user has access to.
+	ReadAllAccessTypeForUser(ctx context.Context, subject string, accessType permission.ObjectType) ([]permission.UserAccess, error)
+}
+
+// UserAccessSpec defines the attributes that can be set when adding a new
+// user access.
+type UserAccessSpec struct {
```

Comment:
I would make a new file under the service package, _types.go_, and put this there.
I'd rename it to `UserAccessArg` and use it as input to `UpdatePermission` as well.